### PR TITLE
rm: fix for -d to match GNU's output #1769

### LINKED ISSUE #1769: rm: -d should match GNU's output
with rm from GNU:
```
$ mkdir a
$ touch a/b
$ rm -d a
rm: cannot remove '/tmp/a': Directory not empty
```
With the Rust implementation:
```
$ mkdir a
$ rm -rf -v a
rm: error: removing '/tmp/a': Directory not empty (os error 39)
```

This will probably requires to update the `show_error` for all binaries (which is fine)

### PR DESCRIPTION
**Pull request by jhscheer**
This fixes #1769
I did not change the `show_error` macro. I'm not sure if that's desired.

There's another PR #1780 for this issue, which is somehow unfeasible since it's approach is to parse the error messages returned from `fs::remove_dir`.
This PR uses the same approach already present in `rmdir`, namely it calls `fs::read_dir` to determine if a directory is empty.

I noticed an anomaly with GNU's output. If asked to remove an empty but non readable directory, it states this directory is not empty. I mimicked this behavior.
```bash
$ mkdir cant_read && chmod -r cant_read
$ rm -d cant_read
rm: cannot remove 'cant_read': Directory not empty
```

There is a related issue #1768 which I also intend to fix. However it would be easier if this would get merged first.

### CODE REVIEW

**Comment by sylvestre** [comment]
Fails on Windows it seems
```

---- test_rm::test_rm_dir_symlink stdout ----
current_directory_resolved: 
mkdir: C:\Users\RUNNER~1\AppData\Local\Temp\.tmpdZguTQ\test_rm_dir_symlink_dir
symlink: C:\Users\RUNNER~1\AppData\Local\Temp\.tmpdZguTQ\test_rm_dir_symlink_dir,C:\Users\RUNNER~1\AppData\Local\Temp\.tmpdZguTQ\test_rm_dir_symlink_link
run: D:\a\coreutils\coreutils\target\x86_64-pc-windows-gnu\debug\coreutils.exe rm test_rm_dir_symlink_link
thread 'test_rm::test_rm_dir_symlink' panicked at 'assertion failed: self.success', tests\common\util.rs:86:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by jhscheer** [comment]
The failing test wants to remove a `symlink_dir`, but doesn't use either the `-d` or the `-r` flag.
@sylvestre What is the desired behavior here?
Should the test change, or the code?

```rust
#[test]
fn test_rm_dir_symlink() {
    let (at, mut ucmd) = at_and_ucmd!();
    let dir = "test_rm_dir_symlink_dir";
    let link = "test_rm_dir_symlink_link";

    at.mkdir(dir);
    at.symlink_dir(dir, link);

    ucmd.arg(link).succeeds();
}
```

**Comment by sylvestre** [comment]
Yeah, it is expected.
try:
```
mkdir a
ln -s a b
rm b
```

**Comment by jhscheer** [comment]
It is expected on unix because "b" is a file.
```bash
$ mkdir a && ln -s a b
$ rmdir b
rmdir: failed to remove 'b': Not a directory
$ rm -v b
removed 'b'
$ rm -dv a
removed directory 'a'
```
On unix it's just [symlink](https://doc.rust-lang.org/std/os/unix/fs/fn.symlink.html) and the `test_rm_dir_symlink` test is not failing on unix.
However, as far as I understand it, on windows there's a distinction between [symlink_dir](https://doc.rust-lang.org/std/os/windows/fs/fn.symlink_dir.html) and [symlink_file](https://doc.rust-lang.org/std/os/windows/fs/fn.symlink_file.html). 

On windows, both [ln.rs](https://github.com/uutils/coreutils/blob/a42e24e088a0ffa38af6549caa304c0cfcf4ac28/src/uu/ln/src/ln.rs#L324) and [mv.rs](https://github.com/uutils/coreutils/blob/bfc9e13f113f11be2762fe0f72d74628db88fd36/src/uu/mv/src/mv.rs#L476) also make a distinction between symlink file/dir.

For `rm.rs` on windows, I think it would make sense to follow ln/mv and to require `-r` or `-d` if `rm` is called to remove a `symlink_dir`.

**Comment by sylvestre** [comment]
@jhscheer Sounds good! I don't know much about Windows TBH

**Comment by jhscheer** [comment]
@sylvestre Me neither, only after the failed test I figured windows makes a distinction between symlink_dir and symlink_file.
Testing was a little cumbersome until I bit the bullet and set up an extra windows VM.

Anyway, this should be good to merge now.

File: src/uu/rm/src/rm.rs [hunk 1]
```diff
@@ -305,16 +305,37 @@ fn remove_dir(path: &Path, options: &Options) -> bool {
...
-            Ok(_) => {
-                if options.verbose {
-                    println!("removed '{}'", path.display());
+        if let Ok(mut read_dir) = fs::read_dir(path) {
+            if options.dir || options.recursive {
+                if read_dir.next().is_none() {
+                    match fs::remove_dir(path) {
+                        Ok(_) => {
+                            if options.verbose {
+                                println!("removed directory '{}'", path.display());
```
Comment (sylvestre) [inline src/uu/rm/src/rm.rs:314]: please add a test to cover these cases

File: src/uu/rm/src/rm.rs [hunk 2]
```diff
@@ -305,16 +305,37 @@ fn remove_dir(path: &Path, options: &Options) -> bool {
...
+                            }
+                        }
+                        Err(e) => {
+                            show_error!("cannot remove '{}': {}", path.display(), e);
```
Comment (sylvestre) [inline src/uu/rm/src/rm.rs:318]: same

**Review by sylvestre — CHANGES_REQUESTED** [review]
Some more tests would be great :)

File: src/uu/rm/src/rm.rs (hunk 1 above)
Comment (jhscheer) [reply src/uu/rm/src/rm.rs:314]: done

File: src/uu/rm/src/rm.rs (hunk 2 above)
Comment (jhscheer) [reply src/uu/rm/src/rm.rs:318]: this is the error for when everything else fails but I don't know how I could provoke it in a test scenario

**Comment by sylvestre** [comment]
don't hesitate to rustfmt the tests (it wasn't correct)

**Comment by jhscheer** [comment]
> don't hesitate to rustfmt the tests (it wasn't correct)

Oh, sorry.
However, I wonder why because I have `rustfmt` on save automatically.

**Pull request merged**